File: src/lod_plan.rs
//! View-dependent LOD decision math (§5/§28).
//!
//! Hosts validate and assemble string mode names; this module owns the
//! exact-vs-aggregate hysteresis decision and the screen-bounded aggregation
//! grid so Python and Node stay bit-identical
//! ([host-parity.md](../../spec/design/host-parity.md)).

/// Cap matching `python/xy/config.py` `MAX_SCREEN_DIM`.
pub const MAX_SCREEN_DIM: i32 = 4096;
/// Floor matching `lod.screen_shape` — avoids zero-size aggregate grids.
pub const MIN_SCREEN_DIM: i32 = 16;
/// Default points-per-cell target (`DENSITY_TARGET_POINTS_PER_CELL`).
pub const DEFAULT_TARGET_PER_CELL: f64 = 16.0;
/// Default drill-exit hysteresis (`DRILL_EXIT_FACTOR`).
pub const DEFAULT_EXIT_FACTOR: f64 = 1.15;

/// Wire/tier mode: ship direct marks (points, candles, …).
pub const MODE_DIRECT: u32 = 0;
/// Wire/tier mode: ship an aggregate (density, buckets, …).
pub const MODE_AGGREGATE: u32 = 1;

/// Chart-agnostic tier decision for a viewport (numeric half of `LodPlan`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct LodPlan {
    pub exact: bool,
    pub mode: u32,
    pub grid_w: i32,
    pub grid_h: i32,
}

/// Clamp a browser/client screen shape to `[MIN_SCREEN_DIM, MAX_SCREEN_DIM]`.
///
/// Matches `lod.screen_shape`: non-positive dimensions floor to the minimum
/// rather than erroring (host validation already rejected non-finite inputs).
pub fn screen_shape(w: i32, h: i32) -> (i32, i32) {
    (
        w.clamp(MIN_SCREEN_DIM, MAX_SCREEN_DIM),
        h.clamp(MIN_SCREEN_DIM, MAX_SCREEN_DIM),
    )
}

/// Hysteresis-guarded drill decision: stay in exact marks until the visible
/// count clearly exceeds the budget again (§5).
pub fn drill_decision(visible: u64, budget: f64, in_drill: bool, exit_factor: f64) -> Option<bool> {
    if !budget.is_finite() || budget <= 0.0 {
        return None;
    }
    if !exit_factor.is_finite() || exit_factor <= 0.0 {
        return None;
    }
    let factor = if in_drill { exit_factor } else { 1.0 };
    let threshold = budget * factor;
    if !threshold.is_finite() {
        return None;
    }
    // Compare in f64; visible fits exactly in f64 up to 2^53.
    Some((visible as f64) <= threshold)
}

/// Keep aggregation grids screen-bounded, but avoid one-pixel bins when the
/// visible count is only barely over the direct budget.
pub fn grid_shape(w: i32, h: i32, visible: u64, target_per_cell: f64) -> Option<(i32, i32)> {
    if !target_per_cell.is_finite() || target_per_cell <= 0.0 {
        return None;
    }
    let (w, h) = screen_shape(w, h);
    let requested = (w as i64).checked_mul(h as i64)?;
    if visible == 0 {
        return Some((w, h));
    }
    let ceil_cells = ((visible as f64) / target_per_cell).ceil();
    if !ceil_cells.is_finite() || ceil_cells < 0.0 {
        return None;
    }
    let cells = ceil_cells as i64;
    let floor = (MIN_SCREEN_DIM as i64) * (MIN_SCREEN_DIM as i64);
    let target = requested.min(cells.max(floor));
    if target >= requested {
        return Some((w, h));
    }
    let scale = (target as f64 / requested as f64).sqrt();
    if !scale.is_finite() {
        return None;
    }
    // Python uses `int(round(...))` — banker's rounding at .5. Match with
    // `round_ties_even` so dual-host grids stay identical.
    let gw = ((w as f64) * scale).round_ties_even() as i32;
    let gh = ((h as f64) * scale).round_ties_even() as i32;
    Some((gw.max(MIN_SCREEN_DIM), gh.max(MIN_SCREEN_DIM)))
}

/// Build the reusable numeric tier decision for a viewport.
pub fn plan(
    visible: u64,
    budget: f64,
    in_drill: bool,
    exit_factor: f64,
    px_w: i32,
    px_h: i32,
    target_per_cell: f64,
) -> Option<LodPlan> {
    let exact = drill_decision(visible, budget, in_drill, exit_factor)?;
    let (grid_w, grid_h) = grid_shape(px_w, px_h, visible, target_per_cell)?;
    Some(LodPlan {
        exact,
        mode: if exact {
            MODE_DIRECT
        } else {
            MODE_AGGREGATE
        },
        grid_w,
        grid_h,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn drill_hysteresis_matches_python_fixtures() {
        assert_eq!(drill_decision(100_000, 200_000.0, false, 1.15), Some(true));
        assert_eq!(drill_decision(350_000, 200_000.0, false, 1.15), Some(false));
        // 225k <= 200k * 1.15 = 230k while drilled → stay exact
        assert_eq!(drill_decision(225_000, 200_000.0, true, 1.15), Some(true));
        // 240k > 230k → exit drill
        assert_eq!(drill_decision(240_000, 200_000.0, true, 1.15), Some(false));
    }

    #[test]
    fn grid_shape_shrinks_when_visible_is_modest() {
        let (gw, gh) = grid_shape(1200, 800, 350_000, 16.0).unwrap();
        assert!(gw < 1200 && gh < 800);
        assert!(gw >= MIN_SCREEN_DIM && gh >= MIN_SCREEN_DIM);
        // Full screen when the target would exceed the pixel grid.
        assert_eq!(grid_shape(64, 48, 0, 16.0), Some((64, 48)));
        assert_eq!(grid_shape(64, 48, 64 * 48 * 16, 16.0), Some((64, 48)));
    }

    #[test]
    fn plan_records_direct_and_aggregate() {
        let direct = plan(100_000, 200_000.0, false, 1.15, 1200, 800, 16.0).unwrap();
        assert!(direct.exact);
        assert_eq!(direct.mode, MODE_DIRECT);

        let aggregate = plan(350_000, 200_000.0, false, 1.15, 1200, 800, 16.0).unwrap();
        assert!(!aggregate.exact);
        assert_eq!(aggregate.mode, MODE_AGGREGATE);
        assert!(aggregate.grid_w < 1200);
    }

    #[test]
    fn screen_shape_clamps_including_non_positive() {
        assert_eq!(screen_shape(0, 10), (16, 16));
        assert_eq!(screen_shape(8, 10_000), (16, MAX_SCREEN_DIM));
    }

    #[test]
    fn rejects_non_finite_policy() {
        assert!(drill_decision(10, f64::NAN, false, 1.15).is_none());
        assert!(grid_shape(64, 48, 10, 0.0).is_none());
    }
}

File: python/xyg/_lod_plan.py
"""LOD tier planning and aggregation grid sizing."""

from __future__ import annotations

import numpy as np

from . import kernels
from ._lod_params import _float_param, _integer_param
from ._lod_types import LodPlan
from ._lod_viewport import ViewportRequest, screen_shape
from .config import DENSITY_TARGET_POINTS_PER_CELL, DRILL_EXIT_FACTOR


def drill_decision(
    visible: int, budget: float, in_drill: bool, exit_factor: float = DRILL_EXIT_FACTOR
) -> bool:
    """The render tier is a function of the *visible* point count (design
    dossier §5), hysteresis-guarded — once drilled down to real points, stay
    until the count clearly exceeds the budget again.

    Decision math lives in Rust (`xy_drill_decision`); this host only coerces
    arguments.
    """
    return bool(kernels.drill_decision(visible, budget, in_drill, exit_factor))


def plan_view_lod(
    request: ViewportRequest,
    visible: object,
    budget: object,
    in_drill: bool,
    *,
    direct_mode: str = "points",
    aggregate_mode: str = "density",
    aggregate_reduction: str = "count",
    target_per_cell: float = DENSITY_TARGET_POINTS_PER_CELL,
    exit_factor: float = DRILL_EXIT_FACTOR,
) -> LodPlan:
    """Build the reusable tier decision for a viewport.

    The exact-vs-aggregate decision is common across tiered chart kinds; the
    representation names differ. Scatter passes `points`/`density`, while
    future histograms or candlesticks can pass `bins`/`ohlc-buckets` without
    reimplementing validation, hysteresis, or screen-bounded grid sizing.
    """
    visible_i = _integer_param(visible, "visible")
    budget_f = _float_param(budget, "LOD budget", min_exclusive=0.0)
    if not isinstance(in_drill, (bool, np.bool_)):
        raise ValueError("in_drill must be True or False")
    for value, label in (
        (direct_mode, "direct_mode"),
        (aggregate_mode, "aggregate_mode"),
        (aggregate_reduction, "aggregate_reduction"),
    ):
        if not isinstance(value, str) or not value:
            raise ValueError(f"{label} must be a non-empty string")
    exit_f = _float_param(exit_factor, "exit_factor", min_exclusive=0.0)
    target_f = _float_param(target_per_cell, "target_per_cell", min_exclusive=0.0)
    # Rust owns the numeric decision; host maps mode ids onto wire strings.
    exact, _mode, gw, gh = kernels.lod_plan(
        visible_i,
        budget_f,
        bool(in_drill),
        exit_factor=exit_f,
        width=request.width,
        height=request.height,
        target_per_cell=target_f,
    )
    return LodPlan(
        mode=direct_mode if exact else aggregate_mode,
        tier="direct" if exact else aggregate_mode,
        visible=visible_i,
        budget=budget_f,
        grid_w=gw,
        grid_h=gh,
        reduction="none" if exact else aggregate_reduction,
        exact=exact,
    )


def grid_shape(
    w: int, h: int, visible: int, target_per_cell: float = DENSITY_TARGET_POINTS_PER_CELL
) -> tuple[int, int]:
    """Keep aggregation grids screen-bounded, but avoid one-pixel bins when
    the visible count is only barely over the direct budget. A few points per
    cell gives smoother drill-out aggregates and smaller updates.

    Host validates/clamps the screen shape; Rust owns the shrink math
    (`xy_lod_grid_shape`).
    """
    w, h = screen_shape(w, h)
    visible_i = _integer_param(visible, "visible")
    target_f = _float_param(target_per_cell, "target_per_cell", min_exclusive=0.0)
    return kernels.lod_grid_shape(w, h, visible_i, target_f)


def local_log_density(
    xs: np.ndarray,
    ys: np.ndarray,
    lo_x: float,
    hi_x: float,
    lo_y: float,
    hi_y: float,
    gw: int,
    gh: int,
) -> np.ndarray:
    """Per-point log-normalized local density in [0,1] — the LUT coordinate
    the client blends during the drill handoff so freshly drilled marks wear
    the aggregate's colormap (never a palette jump)."""
    if len(xs) and hi_x > lo_x and hi_y > lo_y:
        return kernels.local_log_density(xs, ys, lo_x, hi_x, lo_y, hi_y, gw, gh)
    return np.zeros(len(xs), dtype=np.float32)
